test/e2e/build_pipeline_test.go

```diff
@@ -0,0 +1,329 @@
+	}
+	if cond := b.Status.GetCondition(duckv1alpha1.ConditionSucceeded); cond == nil {
+		t.Fatalf("Condition for build %q was nil", buildName)
+	} else if cond.Status != corev1.ConditionTrue {
+		t.Fatalf("Build %q was not successful", buildName)
+	}
+
+	// Update the Configuration with an environment variable, which should trigger a new revision
+	// to be created, but without creating a new build.
+	if err := updateConfigWithEnvVars(clients, names, []corev1.EnvVar{{
+		Name:  "FOO",
+		Value: "bar",
+	}}); err != nil {
+		t.Fatalf("Failed to update config with environment variables: %v", err)
+	}
+
+	nextRevName, err := getNextRevisionName(clients, names)
+	if err != nil {
+		t.Fatalf("Error waiting for next revision to be created: %v", err)
+	}
+	names.Revision = nextRevName
+
+	nextRev, err := clients.ServingClient.Revisions.Get(nextRevName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get latest Revision: %v", err)
+	}
+
+	if diff := cmp.Diff(rev.Spec.BuildRef, nextRev.Spec.BuildRef); diff != "" {
+		t.Fatalf("Unexpected differences in BuildRef: %v", diff)
+	}
+
+	// Update the Configuration's Build with an annotation, which should trigger the creation
+	// of BOTH a Build and a Revision.
+	if err := updateConfigWithBuildAnnotation(clients, names, map[string]string{
+		"testing.knative.dev/foo": "bar",
+	}); err != nil {
+		t.Fatalf("Failed to update config with environment variables: %v", err)
+	}
+
+	nextRevName, err = getNextRevisionName(clients, names)
+	if err != nil {
+		t.Fatalf("Error waiting for next revision to be created: %v", err)
+	}
+	names.Revision = nextRevName
+
+	nextRev, err = clients.ServingClient.Revisions.Get(nextRevName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get latest Revision: %v", err)
+	}
+
+	if diff := cmp.Diff(rev.Spec.BuildRef, nextRev.Spec.BuildRef); diff == "" {
+		t.Fatalf("Got matching BuildRef, wanted different: %#v", rev.Spec.BuildRef)
+	}
+}
+
+func TestBuildAndServePipelineRun(t *testing.T) {
+	clients := Setup(t)
+
+	// Add test case specific name to its own logger.
+	logger := logging.GetContextLogger("TestBuildAndServePipelineRun")
+
+	logger.Infof("Creating Pipeline and Task for the build with PipelineRun")
+	task := &pipelinev1alpha1.Task{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: test.ServingNamespace,
+			Name:      "hello-task",
+		},
+		Spec: pipelinev1alpha1.TaskSpec{
+			Steps: []corev1.Container{{
+				Name:  "foo",
+				Image: "busybox",
+				Args:  []string{"echo", "hellow"},
+			}},
+		},
+	}
+	pipeline := &pipelinev1alpha1.Pipeline{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: test.ServingNamespace,
+			Name:      "hello-pipeline",
+		},
+		Spec: pipelinev1alpha1.PipelineSpec{
+			Tasks: []pipelinev1alpha1.PipelineTask{{
+				Name: "hello-pipeline-hello-task",
+				TaskRef: pipelinev1alpha1.TaskRef{
+					Name: "hello-task",
+				},
+			}},
+		},
+	}
+	if _, err := clients.PipelineClient.Task.Create(task); err != nil {
```

Comment:
It could be nice to just inline `task` here, like:

```
if _, err := clients.PipelineClient.Task.Create(&pipelinev1alpha1.Task{
  ObjectMeta: metav1.ObjectMeta{
    ...
  },
  Spec: {...},
}); err != nil {
  t.Fatalf(...)
}
```

This makes it clear upon scanning that the `Task` isn't referenced anywhere else, and is just created so that it exists.

Likewise with `pipeline`, which you don't need since you only pass around the name.